### PR TITLE
AddMethodCallBasedStrictParamTypeRector: Support changing protected methods in final classes

### PR DESCRIPTION
**Pull request by staabm**

### CODE REVIEW

**Comment by staabm** [comment]
~~I think the failling build is not related to the PR~~ seems like a temporary problem which no longer reproduces

File: rules/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector.php [hunk 1]
```diff
@@ -88,7 +89,13 @@ public function refactor(Node $node): ?Node
                 continue;
             }
 
-            if (! $method->isPrivate()) {
+            $isPrivate =
+                ($node->isFinal()
+                && $node->extends === null
```
Comment (samsonasik) [inline rules/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector.php:94]: `$node->implements` empty should be checked as well, also probably used by/from trait.

service from Parent override guard or FamilyTree services may be can be used for this case

Comment (staabm) [reply rules/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector.php:94]: good point. added a test-case and fixed it

Comment (staabm) [reply rules/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector.php:94]: I did not use the suggested services, as I think pure [AST based decision are more reliable right now](https://github.com/rectorphp/rector/issues/8093)

Comment (samsonasik) [reply rules/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector.php:94]: Ok, lets use this as initial support and then we can improve later with verify parent FullyQualified->toString() is autoloaded

**Review by samsonasik — APPROVED** [review]

**Comment by samsonasik** [comment]
Thank you @staabm

**Pull request merged**